src/cpp/fastdds/xtypes/dynamic_types/DynamicDataImpl.cpp

```diff
@@ -0,0 +1,7218 @@
+{
+    return set_value<TK_FLOAT128>(id, value);
+}
+
+ReturnCode_t DynamicDataImpl::set_char8_value(
+        MemberId id,
+        char value) noexcept
+{
+    return set_value<TK_CHAR8>(id, value);
+}
+
+ReturnCode_t DynamicDataImpl::set_char16_value(
+        MemberId id,
+        wchar_t value) noexcept
+{
+    return set_value<TK_CHAR16>(id, value);
+}
+
+ReturnCode_t DynamicDataImpl::set_byte_value(
+        MemberId id,
+        eprosima::fastrtps::rtps::octet value) noexcept
+{
+    return set_value<TK_BYTE>(id, value);
+}
+
+ReturnCode_t DynamicDataImpl::set_boolean_value(
+        MemberId id,
+        bool value) noexcept
+{
+    return set_value<TK_BOOLEAN>(id, value);
+}
+
+ReturnCode_t DynamicDataImpl::set_string_value(
+        MemberId id,
+        const std::string& value) noexcept
+{
+    if (TK_STRING8 == enclosing_type_->get_kind())
+    {
+        auto bound = enclosing_type_->get_descriptor().bound().at(0);
+        if (0 != bound && value.length() > bound)
+        {
+            EPROSIMA_LOG_ERROR(DYN_TYPES,
+                    "Error setting string value. The given string is greater than the length limit.");
+            return RETCODE_BAD_PARAMETER;
+        }
+    }
+
+    return set_value<TK_STRING8>(id, value);
+}
+
+ReturnCode_t DynamicDataImpl::set_wstring_value(
+        MemberId id,
+        const std::wstring& value) noexcept
+{
+    if (TK_STRING16 == enclosing_type_->get_kind())
+    {
+        auto bound = enclosing_type_->get_descriptor().bound().at(0);
+        if (0 != bound && value.length() > bound)
+        {
+            EPROSIMA_LOG_ERROR(DYN_TYPES,
+                    "Error setting string value. The given string is greater than the length limit.");
+            return RETCODE_BAD_PARAMETER;
+        }
+    }
+
+    return set_value<TK_STRING16>(id, value);
+}
+
+//}}}
+
+ReturnCode_t DynamicDataImpl::set_complex_value(
+        MemberId id,
+        traits<DynamicData>::ref_type value) noexcept
+{
+    ReturnCode_t ret_value = RETCODE_BAD_PARAMETER;
+
+    if (value && id != MEMBER_ID_INVALID)
+    {
+        TypeKind type_kind = enclosing_type_->get_kind();
+
+        if (TK_ANNOTATION == type_kind ||
+                TK_BITSET == type_kind ||
+                TK_STRUCTURE == type_kind ||
+                TK_UNION == type_kind)
+        {
+            if (TK_UNION != type_kind || 0 != id)
+            {
+                auto it = value_.find(id);
+                if (it != value_.end())
+                {
+                    auto data = std::static_pointer_cast<DynamicDataImpl>(it->second);
+
+                    if (data->type_->equals(value->type()))
+                    {
+                        value_.erase(it);
+                        value_.emplace(id, value->clone());
+                        return RETCODE_OK;
+                    }
+                }
+
+                if (RETCODE_OK == ret_value && TK_UNION == type_kind && 0 != id) // Set new discriminator.
```

Comment:
The last check `0 != id` is going to be always true because the previous check ensures that if TK_UNION and id 0 coincides, the program flow does not come into this code.